### DISCUSSION TITLE
Missing text after redaction

### DISCUSSION THREAD

**Discussion by NoraishaYusuf** [pymupdf/PyMuPDF#867] [Looking for help] [answered]
I'm currently redacting sensitive information in resumes. 
There are two major issues i'm facing as follows:

1. I found chunk of texts & headings, and sometimes photos missing for several of the redacted resumes. 
For one of the resumes, these were the warnings thrown (not in the right order):
_mupdf: unknown colorspace: R26
mupdf: unknown colorspace: R28
mupdf: syntax error in content stream
mupdf: unknown keyword: 'EI'
mupdf: unknown colorspace: R27
mupdf: unknown keyword: 'kd'
mupdf: array not closed before end of file_

[](url)
I have attached the relevant part of my codes for reference. 
![image](https://user-images.githubusercontent.com/43775019/105746089-f91c6000-5f79-11eb-89f4-f0270e9eee05.png)

2. The redacted word overlaps with words located above and below it (as shown in the images)

![image](https://user-images.githubusercontent.com/43775019/105747214-78f6fa00-5f7b-11eb-9505-b7a5a66f6031.png)
![image](https://user-images.githubusercontent.com/43775019/105747341-a04dc700-5f7b-11eb-9461-19c86f36db13.png)

I appreciate it very much if anyone can help me solve these two issues.

**Comment by JorjMcKie** [comment, marked as answer]
# Problem 1:
Hard to tell without looking at that file (probably confidential anyway). But there arethings like damaged PDFs ...
You could try cleaning the file / the page before processing to reveal / remove any errors.
* cleaning the file e.g. by ``mutool clean -gggsc file.pdf``
* cleaning the page inside the script: ``page.clean_contents(sanitize=True)``

# Problem 2:
MuPDF normally uses the full font-defined line height when identifying the hits of search. If the PDF is made with smaller distances between lines, then adjacent lines may overlap somewhat. The redaction logic of MuPDF in turn removes every character **_overlaping_** the redaction rectangle - the result of this is what you saw.
* set a global parameter that reduces character bbox heights to the fontsize value: `fitz.TOOLS.set_small_glyph_heights(True)` - right after importing fitz.

**Reply by mani2106** [reply]
Does the `clean_contents` method exist for a `Page` object as per fitz version `1.17.0`?

**Reply by JorjMcKie** [reply]
Yes! Should have been callled `page.cleanContents()` then.
